File: Python/Binary Search/Search a 2D matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        left = 0
        m,n = len(matrix),len(matrix[0])
        right = m*n - 1

        while(left <= right) :
            mid = (left+right)//2

            if matrix[mid//n][mid%n] == target :
                return True
            elif matrix[mid//n][mid%n] < target:
                left = mid +1
            else:
                right = mid -1

        return False

File: C++/Binary Search/Search a 2D matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {

        int left =0 , m = matrix.size() , n= matrix[0].size();
        int right = m*n-1;

        while(left <= right) {
            int mid = left + (right-left)/2;
            // cout << "mid-" << mid << "-left -" << left << "-right-" << right << endl;
            if(matrix[mid/n][mid%n] == target){
                return true;
            }
            else if(matrix[mid/n][mid%n] < target){
                left = mid + 1;
            }else{
                right = mid - 1;
            }

        }

        return false;


    }
};

//Note always divide by number of columns to get the matrix indices